lambdas/initialise-ipv-session/src/main/java/uk/gov/di/ipv/core/initialiseipvsession/InitialiseIpvSessionHandler.java

```diff
@@ -227,6 +251,76 @@ public APIGatewayProxyResponseEvent handleRequest(
         }
     }
 
+    @Tracing
+    private void validateAndStoreHMRCInheritedIdentity(
+            String userId, JWTClaimsSet claimsSet, IpvSessionItem ipvSessionItem)
+            throws RecoverableJarValidationException, ParseException {
+        JarClaims claims;
+        try {
+            claims =
+                    OBJECT_MAPPER.convertValue(
+                            claimsSet.getJSONObjectClaim(CLAIMS_CLAIM), JarClaims.class);
+        } catch (IllegalArgumentException | ParseException e) {
+            throw buildRecoverableException(
+                    OAuth2Error.INVALID_REQUEST_OBJECT.setDescription("Unable to process claims"),
+                    claimsSet);
+        }
+        if (claims == null) {
+            return;
+        }
+
+        var userInfo = claims.userInfo();
+        if (userInfo == null) {
+            return;
+        }
+
+        var inheritedIdentityJwtClaim = userInfo.inheritedIdentityClaim();
+        if (inheritedIdentityJwtClaim == null) {
+            return;
+        }
+
+        SignedJWT signedInheritedIdentityJWT;
+        try {
+            signedInheritedIdentityJWT = SignedJWT.parse(inheritedIdentityJwtClaim.value().get(0));
+            CriConfig inheritedIdentityCriConfig = configService.getCriConfig(HMRC_MIGRATION_CRI);
+
+            verifiableCredentialJwtValidator.validate(
+                    signedInheritedIdentityJWT, inheritedIdentityCriConfig, userId);
+            LogHelper.logMessage(Level.INFO, "Migration VC successfully validated");
+        } catch (VerifiableCredentialException | ParseException e) {
+            throw buildRecoverableException(
+                    OAuth2Error.INVALID_REQUEST_OBJECT
+                            .setDescription("Inherited identity JWT failed to validate - ")
+                            .appendDescription(e.getMessage()),
+                    claimsSet);
+        }
+
+        try {
+            verifiableCredentialService.persistUserCredentials(
+                    signedInheritedIdentityJWT, HMRC_MIGRATION_CRI, userId);
+            ipvSessionItem.setVcReceivedThisSession(
+                    List.of(signedInheritedIdentityJWT.serialize()));
```

Comment:
I wonder if we should add an `addVcReceivedThisSession` method to `IpvSessionItem` - I think it might simplify the code in the other use-case as well...